src/lib.rs

```diff
@@ -130,40 +135,48 @@ unsafe fn console_on_any(fds: &[STD_HANDLE]) -> bool {
 
 /// Returns true if there is an MSYS tty on the given handle.
 #[cfg(windows)]
-unsafe fn msys_tty_on(fd: STD_HANDLE) -> bool {
+fn msys_tty_on(handle: HANDLE) -> bool {
     use std::ffi::c_void;
-    use windows_sys::Win32::{
-        Foundation::MAX_PATH,
-        Storage::FileSystem::{FileNameInfo, GetFileInformationByHandleEx},
-        System::Console::GetStdHandle,
+    use windows_sys::Win32::Storage::FileSystem::{
+        FileNameInfo, GetFileInformationByHandleEx, FILE_NAME_INFO,
     };
 
-    /// Mirrors windows_sys::Win32::Storage::FileSystem::FILE_NAME_INFO, giving
-    /// it a fixed length that we can stack allocate
-    #[repr(C)]
-    #[allow(non_snake_case)]
-    struct FILE_NAME_INFO {
-        FileNameLength: u32,
-        FileName: [u16; MAX_PATH as usize],
-    }
+    // Determine the length of the file name.
     let mut name_info = FILE_NAME_INFO {
         FileNameLength: 0,
-        FileName: [0; MAX_PATH as usize],
+        FileName: [0; 1],
+    };
+    unsafe {
+        GetFileInformationByHandleEx(
+            handle,
+            FileNameInfo,
+            &mut name_info as *mut _ as *mut c_void,
+            std::mem::size_of::<FILE_NAME_INFO>() as u32,
+        );
+    }
+    // The length is in bytes, so check that it is divisible by two.
+    if name_info.FileNameLength == 0 || name_info.FileNameLength % 2 != 0 {
+        return false;
+    }
+    let required_length = name_info.FileNameLength as usize / 2;
+
+    // Allocate a buffer of the required size in bytes plus 2 array element
+    // (4 bytes) for the FileNameLength field.
+    let mut name_info = vec![0u16; required_length + 2];
```

Comment:
This doesn't ensure that the buffer has the required alignment for a `FILE_NAME_INFO` to be stored in it. Would it work to allocate a slightly larger buffer and use [`align_to`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.align_to)?

Reply:
I believe this function serves a slightly different purpose and `std::alloc::Layout` should be used instead, but it isn't trivial. I have reverted most of the changes to use stack allocated `FILE_NAME_INFO` with `MAX_PATH` fixed size. This fixes panic and the case is already very rare, so personally I don't mind false negatives. What do you think?